src/datasets/features/image.py

```diff
@@ -0,0 +1,287 @@
+
+        Args:
+            value (:obj:`str`, :obj:`np.ndarray`, :obj:`PIL.Image.Image` or :obj:`dict`): Data passed as input to Image feature.
+
+        Returns:
+            :obj:`dict`
+        """
+        if config.PIL_AVAILABLE:
+            import PIL.Image
+        else:
+            raise ImportError("To support encoding images, please install 'Pillow'.")
+
+        if isinstance(value, str):
+            return {"path": value, "bytes": None}
+        elif isinstance(value, np.ndarray):
+            image = PIL.Image.fromarray(value.astype(np.uint8))
+            return {"path": None, "bytes": image_to_bytes(image)}
+        elif isinstance(value, PIL.Image.Image):
+            if hasattr(value, "filename") and value.filename != "":
+                return {"path": value.filename, "bytes": None}
+            else:
+                return {"path": None, "bytes": image_to_bytes(value)}
+        else:
+            return value
+
+    def decode_example(self, value):
+        """Decode example image file into image data.
+
+        Args:
+            value (obj:`str` or :obj:`dict`): a string with the absolute image file path, a dictionary with
+                keys:
+                - path: String with absolute or relative image file path.
+                - bytes: The bytes of the image file.
+
+        Returns:
+            :obj:`PIL.Image.Image`
+        """
+        if config.PIL_AVAILABLE:
+            import PIL.Image
+        else:
+            raise ImportError("To support decoding images, please install 'Pillow'.")
+
+        if isinstance(value, str):
+            path, bytes_ = value, None
+        else:
+            path, bytes_ = value["path"], value["bytes"]
+
+        if bytes_ is None:
+            if isinstance(path, str):
+                if is_local_path(path):
+                    image = PIL.Image.open(path)
+                else:
+                    with xopen(path, "rb") as f:
+                        bytes_ = BytesIO(f.read())
+                    image = PIL.Image.open(bytes_)
+        else:
+            image = PIL.Image.open(BytesIO(bytes_))
+        return image
+
+
+def list_image_compression_formats():
+    if config.PIL_AVAILABLE:
+        import PIL.Image
+    else:
+        raise ImportError("To support encoding images, please install 'Pillow'.")
+
+    global _IMAGE_COMPRESSION_FORMATS
+    if _IMAGE_COMPRESSION_FORMATS is None:
+        PIL.Image.init()
+        _IMAGE_COMPRESSION_FORMATS = list(set(PIL.Image.OPEN.keys()) & set(PIL.Image.SAVE.keys()))
+    return _IMAGE_COMPRESSION_FORMATS
+
+
+def image_to_bytes(image: "PIL.Image.Image") -> bytes:
+    """Convert a PIL Image object to bytes using native compression if possible, otherwise use PNG compression."""
+    buffer = BytesIO()
+    format = image.format if image.format in list_image_compression_formats() else "PNG"
+    image.save(buffer, format=format)
+    return buffer.getvalue()
+
+
+def objects_to_list_of_image_dicts(objs):
```

Comment:
```suggestion
def objects_to_list_of_image_dicts(objs: Union[List[str], List[dict], List[np.ndarray], List["PIL.Image.Image"]]) -> List[dict]:
```

Reply:
Waiting to resolve https://github.com/huggingface/datasets/pull/3163#discussion_r760089258 first